lib/logstash/filters/geoip.rb

```diff
@@ -108,10 +124,23 @@ def tag_unsuccessful_lookup(event)
     @tag_on_failure.each{|tag| event.tag(tag)}
   end
 
+  def setup_target_field
+    if ecs_compatibility == :disabled
+      @target ||= 'geoip'
+    else
+      auto_target = @source[0...-4] if @source.end_with?('[ip]') and @source.length > 4
+      @target ||= auto_target || fail(LogStash::ConfigurationError, "GeoIP Filter in ECS-Compatiblity mode "\
+                                            "requires a `target` when `source` is not an `ip` sub-field, eg. [client][ip]")
+
+      normalized_target = (@target.start_with?('[') and @target.end_with?(']'))? @target[1...-1] : @target
+      logger.warn("ECS expect `target` value in #{ECS_TARGET_FIELD}") unless ECS_TARGET_FIELD.include?(normalized_target)
+    end
+  end
```

Comment:
There is a lot going on with conditional assignments, which could be made a lot simpler by extracting `auto_target_from_source` to its own method.

A couple separate points:

1. We tend to avoid the `and` and `or` keywords in Ruby conditionals (favoring their symbolic counterparts `&&` and `||`) because they have surprising behaviour with regard to precedence (more: https://avdi.codes/how-to-use-rubys-english-andor-operators-without-going-nuts/).
2. I'm wary of just stripping open-and-close brackets to normalize the field reference, because nested fields end up with a funky syntax that would be hard to replicate if/when ECS adds a host-like field that is _not_ at the top-level. Using an anchored-at-both-ends gsub like below will allow us to strip the bracket syntax from top-level fields while leaving nested fields alone.

```suggestion
  def setup_target_field
    if ecs_compatibility == :disabled
      @target ||= 'geoip'
    else
      @target ||= auto_target_from_source!

      # normalize top-level fields to not be bracket-wrapped
      normalized_target = @target.gsub(/\A\[([^\[\]]+)\]\z/,'\1')

      logger.warn("ECS expect `target` value `#{normalized_target}` in #{ECS_TARGET_FIELD}") unless ECS_TARGET_FIELD.include?(normalized_target)
    end
  end

  def auto_target_from_source!
    return @source[0...-4] if @source.end_with?('[ip]') && @source.length > 4
      
    fail(LogStash::ConfigurationError, "GeoIP Filter in ECS-Compatiblity mode "\
                                       "requires a `target` when `source` is not an `ip` sub-field, eg. [client][ip]")
  end
```

Reply:
1 :exploding_head: